Add tutorial code (#1)
sisyphusSmiling/attachments-tutorial: README.mdx

build something entirely new.

Composability on Flow, especially with attachments is permissionless because not only can you build with existing 
components, but no one can stop you from doing so!

</Callout>

Permissionless composability is huge because, as builders, we can build on existing constructs for existing 
communities and be sure that the original developer won't rug our access - AKA platform risk.

For an example of the negative case, just look at how big tech companies locked down their API in the early days.

Less duplicate code, easier to bootstrap attention and adoption, and users benefit from the utility you built into 
something they already presumably own and love - win-win-win!

# Define our Base NFT - KittyVerse

## NonFungibleToken Standard

- Building on the existing standard
- Encouraged to also implement MetadataViews and ViewResolver contract interfaces, but keeping things simple
- We're defining the KittyVerse NFT for the purposes of this tutorial, but just as easily could've used an existing NFT 
project since we're building with attachments
- If totally new to this, take a look at NonFungibleToken tutorials - TODO: Link these

## Contract Overview

> - Contract interface implementation

Reviewer: I would suggest to give an overview about the design choices and give a disclaimer that it is up to the developer to use attachment for other purposes and they can have there own design choices but we recommend to follow the best practices of writing the smart contracts in cadence.

    - Events

>     - Nested resource definitions

Reviewer: Pointer - Maybe it will be good to add that with attachments it is something not possible to do so or if possible then what were the drawbacks of the same so a developer can easily quantify the benefits of the attachments.

- Attribute overview
- Note on `unsafeRandom()`
- Opportunities to extend the contract

## Deploy KittyVerse contract

- Deploy NonFungibleToken to `0x01`
- Deploy KittyVerse to `0x02`

## Setup account & Mint KittyVerse NFT

> - send `tx_01_setup_kittyverse_account.cdc` signing with `0x02`

Reviewer: I am assuming that you are going to broaden the transactions as well. Like what they are doing and what are the important nuances in the transactions and scripts that a developer should keep remembering it. Maybe good to add some links of transaction-related documentation

Author: Yes definitely will expand on the details of the transaction similar to how other current tutorials walk through the details of each transaction.

- send `tx_02_mint_kittyverse_nft.cdc` signing with `0x02`
- send `tx_02_mint_kittyverse_nft.cdc` again signing with `0x02`
- query NFT.id - execute `script_01_get_kittyverse_ids.cdc` providing `0x02` as argument
    - see two unique NFT IDS - take note of them, next let's prep them some hats

# Define our Attachments - KittyHats

## Attachments

- What are they?
    - In Cadence
    - Simple mental model

>         - "Like glue between resources"

Reviewer: Don't you think we need more detailed documentation here? Maybe a link to a more detailed version of the documentation.

Author: Yes definitely, though it's just an outline for now. I wanted to get some eyes on where things are in code before diving too deep on documentation.

        - ![TODO - Diagram](./attachment-diagram.png)
- Why do they matter?
    - Can create attachments for anything without needing original developer's:
        - Participation
        - Permission

## Contract Overview

### NFT

- Definitions

### HatAttachment

- Common misconception: The attachment is the thing
    - Correction: Attachments glue the thing(s)
        - As mentioned previously, attachments are like glue between resources
    - As of now, attachments can be fully referenced anywhere the NFT can be referenced - this can complicate access 
    control

>         - We made `removeHat()` `access(contract)` and callable in Collection only

Reviewer: Maybe a link for access modifiers documentation

## Setup account & Mint KittyHats NFT

- send `tx_03_setup_kittyhats_account.cdc` signing with `0x02`
- query NFT.id - `script_03_get_kittyhats_ids_and_names.cdc` providing `0x02` as argument
    - see one id and note it for the next step

# Put hats on cats

- send `tx_05_attach_hat_to_cat.cdc` signing with `0x02`
    - provid KittyHats.NFT.id & KittyVerse.NFT.id of your choice
- query cats & hats - `script_04_get_kittyverse_names_and_hat` providing `0x02` as argument
    - see the KittyVerse.NFT names - one should have your hat name and the other should have nil - Your cat has a hat!
- query KittyHats NFT IDs - `script_03_get_kittyhats_ids_and_names.cdc` providing `0x02` as argument
    - see that your hat is no longer in your Collection

# Remove the hat

- send `tx_06_remove_hat_from_cat.cdc` signing with `0x02` and providing your KittyVerse NFT id name
- query cats & hats - `script_04_get_kittyverse_names_and_hat` providing `0x02` as argument
    - see that your cat's hat has been removed
- query KittyHats NFT IDs - `script_03_get_kittyhats_ids_and_names.cdc` providing `0x02` as argument
    - see that your hat is back in your Collection

# Extra Credit

> 

Reviewer: Don't you go to explain how the attachment works or whatnot from the code .. it looks like a walkthrough? Or am I missing something?

Author: I mentioned it above, but to clarify here - as it stands now it's just an outline of points to expand on atm. Was hoping to validate the code and the idea before investing more time in the documentation.

Reviewer: Okay cool make sense

# TODO: Expand on outline

> :warning: This guidance is in prep for a Playground tutorial, but attachments must first be enabled on Playground for
this guidance to make sense.

# KittyVerse + KittyHats

To showcase how attachments unlock permissionless composability on Flow, this tutorial will walk you though making your
own cat-inspired NFT collection and a wardrobe of hats to go along with them. Let's put some hats on some cats!

## What is composability?

Before digging in, let's talk about what we mean by **_composability_**.

<Callout type="info">

Composability is the ability for existing components in one construction to be reused and even extended for another 
purpose. <br/>

Think of composable architectures like [building with legos](https://twitter.com/cdixon/status/1450462593232089088?s=20) 
- you build something you need as a block or set of blocks and someone else can come along and use those same blocks to 

> build something entirely new.

Reviewer 2: Maybe we shine a light on strengths Cadence specifically has for composability namely that contracts can compose other contracts and also load them dynamically on-demand. This makes working with contracts more similar to libraries than any other smart contract language and this education will help them see things more flexibly

Author: Good idea! Do we have docs that expand on this?

Reviewer 2: Borrowing contracts shows it best: https://developers.flow.com/cadence/language/contracts#borrowing-a-deployed-contract

> Composability on Flow, especially with attachments is permissionless because not only can you build with existing 
> components, but no one can stop you from doing so!

Reviewer 2: ```suggestion
Composability on Flow, especially with attachments is permissionless because not only can you build with existing 
components, but no one can stop you from doing so! Hence the term permissionless!
```

</Callout>

> Permissionless composability is huge because, as builders, we can build on existing constructs for existing 
> communities and be sure that the original developer won't rug our access - AKA platform risk.
> 
> For an example of the negative case, just look at how big tech companies locked down their API in the early days.
> 
> Less duplicate code, easier to bootstrap attention and adoption, and users benefit from the utility you built into 
> something they already presumably own and love - win-win-win!

Reviewer 2: ```suggestion
Permissionless composability is one of the unique characteristics of decentralized app development because, as builders, we can freely compose richer or derivative functionalities on top of any existing contracts from other applications or communities. Contracts that you deploy which interact with other contracts are not subject to internal changes or control by the contract you composed it upon. 

Less duplicate code, easier to bootstrap attention and adoption, and users benefit from the utility you built into 
something they already presumably own and love - win-win-win!
```

I struggle a little with the 'original developer won't rug our access' since for Web3 newcomers it might not be enough for them to make a connection. We may want to drop the point about platform risk - it's still the case that if you build to Flow you can't easily switch platforms whereas the same doesn't apply to EVM compatible equivalents. 

I'm also not sure there's enough strength in the reference to big tech APIs for the suggested wording. GCP copied AWS in terms of APIs to smooth migration, so in a sense the locked down aspects of the cloud provider APIs is their strength since they are so ubiquitously supported/similar

Author: > it's still the case that if you build to Flow you can't easily switch platforms whereas the same doesn't apply to EVM compatible equivalents.

Definitely a valid point. By "platform" here I mean something like Twitter's API when they proclaimed to be a protocol, not necessarily the the blockchain itself. The ground may at any point fall out from under you if the chain ceases to exist, but, taking the chain's existence as a given, I'd say the immutability guarantees of smart contracts provide a risk-reduced platform in comparison to privately owned and managed APIs. Case in point being the aforementioned Twitter API.

Regardless, your verbiage gets the point across more clearly for sure. Thanks for the rephrase!

# Define our Base NFT - KittyVerse

## NonFungibleToken Standard

- Building on the existing standard
- Encouraged to also implement MetadataViews and ViewResolver contract interfaces, but keeping things simple
- We're defining the KittyVerse NFT for the purposes of this tutorial, but just as easily could've used an existing NFT 
project since we're building with attachments
- If totally new to this, take a look at NonFungibleToken tutorials - TODO: Link these

## Contract Overview

- Contract interface implementation
    - Events
    - Nested resource definitions
- Attribute overview

> - Note on `unsafeRandom()`

Reviewer 2: Maybe leave this out for now since we're about to launch SafeRandom in this sprint cycle

- Deploy NonFungibleToken to `0x01`
- Deploy KittyVerse to `0x02`

## Setup account & Mint KittyVerse NFT

- send `tx_01_setup_kittyverse_account.cdc` signing with `0x02`
- send `tx_02_mint_kittyverse_nft.cdc` signing with `0x02`
- send `tx_02_mint_kittyverse_nft.cdc` again signing with `0x02`
- query NFT.id - execute `script_01_get_kittyverse_ids.cdc` providing `0x02` as argument
    - see two unique NFT IDS - take note of them, next let's prep them some hats

# Define our Attachments - KittyHats

## Attachments

- What are they?
    - In Cadence
    - Simple mental model
        - "Like glue between resources"
        - ![KittyVerse & KittyHats attached](./diagrams/cat_with_hat.png)
- Why do they matter?
    - Can create attachments for anything without needing original developer's:
        - Participation
        - Permission

### Overview

- Conceptual
    - Attachments are a special type that can be added to resources (doc link needed)
    - You might be tempted to simply define all your functionality in the attachment, but you'll lose the existence and
      uniqueness guarantees inherent to resources as well as make access control significantly harder for yourself

>         - As mentioned earlier, attachments are best thought of as glue between resources

Reviewer 2: Maybe we should be explicit and also say that they aren't Resources and shouldn't be thought of that way? Maybe also we should explain that it's utility is not for use when within ones own contracts, but when other contracts want to interact and contribute something new to an existing Resource

Author: > Maybe we should be explicit and also say that they aren't Resources and shouldn't be thought of that way?

Agreed, worth noting that fact

> Maybe also we should explain that it's utility is not for use when within ones own contracts, but when other contracts want to interact and contribute something new to an existing Resource

IDK if we need to be that prescriptive about their use. It's not clear to me that it's necessarily an anti-pattern to develop with attachments within contracts you own. Is there a reason you think it's bad practice?

Reviewer: It would be better not to have the attachment at the very first place. until I know this attachment would be useful to some x-audience. So it would be easy to manage and less error-prone. While we can add attachments at any time during upgrades, isn't it?

Reviewer 2: > It's not clear to me that it's necessarily an anti-pattern to develop with attachments within contracts you own. Is there a reason you think it's bad practice

Not entirely sure myself, it's something that @joshuahannan spoke to more. He argued before that if it's your contract you don't need attachments since you can explicitly design the functionality into the main contract. Maybe we should make sure the documentation makes it clear that providing support for attachments in your Resources is a key way to enable composability from unknown future sources because it signals that the project is open to collaboration/contributions

Author: Definitely don't need attachments if it's your contract. That said, I don't know that we need to discourage that approach.

For instance, let's say I want to create a generic gaming NFT. I might create a base NFT and define my in-game resources as attachments. This means the access pattern across all of the in-game resources are consistent - for those that I defined and others that emerge in the future. I envision this being closest to token-bound accounts.

But I'll add in verbiage around the fact that attachments aren't necessary if you control the contracts.

I do think we can allow for some emergent patterns where an approach isn't apparently harmful, so hesitant to be too prescriptive here.

            - Define an attachment that can hold some resource, designed to be added to a base resource
    - Note that you need access to the base resource to add an attachment to it
        - So to add an attachment to an NFT, you need to first withdraw that NFT from its Collection
- Syntax
    - Accessing an attachment
    ```cadence
    let attachmentRef: &KittyHats.HatAttachment? = nft[KittyHats.HatAttachment]
    ```
    - Adding an attachment
    ```cadence
    access(self) fun addAttachment(toNFT: @KittyVerse.NFT): @KittyVerse.NFT {
        // If attachment already exists, return
        if toNFT[HatAttachment] != nil {
            return <-toNFT
        }
        // Otherwise, add the attachment
        return <- attach HatAttachment() to <- toNFT
    }
    // Otherwise, add the attachment
    return <- attach HatAttachment() to <- toNFT
    ```
- Gotchas
    - Accessing attachments of a given type requires a static type declaration
    - Attempting to add an attachment to a resource when one already exists will result in a runtime error
        - This is why we check if an attachment already exists before adding one in the example above

>         - You'll want to lean on simple helper methods to conditionally add attachments within your contracts

Reviewer 2: Maybe leave Gotcha's to a later section where we summarize what was done in the tutorial and add that color then?

Author: Good idea. I can move it to the latter part for reflection on aspects of the design choices

- Great for

>     - Building on existing resources

Reviewer 2: ```suggestion
    - Building on existing resources in other contracts 
```

# TODO: Expand on outline

> :warning: This guidance is in prep for a Playground tutorial, but attachments must first be enabled on Playground for
this guidance to make sense.

# KittyVerse + KittyHats

> 

Reviewer: Do you think it would be good to add the reference of kittyVerse & kittyHats repo ? or a small context about them ?

## What is composability?

Before digging in, let's talk about what we mean by **_composability_**.

<Callout type="info">

Composability is the ability for existing components in one construction to be reused and even extended for another 
purpose. <br/>

Think of composable architectures like [building with legos](https://twitter.com/cdixon/status/1450462593232089088?s=20) 
- you build something you need as a block or set of blocks and someone else can come along and use those same blocks to 
build something entirely new.<br/>

Composability on Flow, especially with attachments, is permissionless because not only can you build with existing 
components, but no one can stop you from doing so! Hence the term permissionless!

</Callout>

Permissionless composability is one of the unique characteristics of decentralized app development because, as builders,
we can freely compose richer or derivative functionalities on top of any existing contracts from other applications or
communities. Contracts that you deploy which interact with other contracts are not subject to internal changes or
control by the contract you composed it upon. 

> Cadence is uniquely suited for composable architectures because Cadence contracts can compose other contracts. As a

Reviewer: I don't know what is the right way to put it but this line seems odd to me ` because Cadence contracts can compose other contracts`.  Maybe re-write like this `Cadence excels in facilitating composable architectures due to its ability to compose other contracts through Cadence contracts.`

        - Participation
        - Permission

### Overview

- Conceptual
    - Attachments are a special type that can be added to resources (doc link needed)
    - You might be tempted to simply define all your functionality in the attachment, but it's important to know that
      attachments are not resources.
        - You'll lose the existence and uniqueness guarantees inherent to resources as well as make access control
          significantly harder for yourself
        - As mentioned earlier, attachments are best thought of as glue between resources
            - Define an attachment that can hold some resource, designed to be added to a base resource
    - Note that you need access to the base resource to add an attachment to it
        - So to add an attachment to an NFT, you need to first withdraw that NFT from its Collection
- Syntax
    - Accessing an attachment
    ```cadence
    let attachmentRef: &KittyHats.HatAttachment? = nft[KittyHats.HatAttachment]
    ```
    - Adding an attachment
    ```cadence
    access(self) fun addAttachment(toNFT: @KittyVerse.NFT): @KittyVerse.NFT {
        // If attachment already exists, return

>         if toNFT[HatAttachment] != nil {

Reviewer: My assumption is that you are going to break it down this little bit. However, it is a readable code but newcomers may not understand the syntax `toNFT[HatAttachment]` as it only apply to attachments.

Author: Yes, I'll add more context and a general case before introducing contract associated verbiage.

            return <-toNFT
        }
        // Otherwise, add the attachment
        return <- attach HatAttachment() to <- toNFT
    }
    // Otherwise, add the attachment
    return <- attach HatAttachment() to <- toNFT
    ```
   
## Contract Overview

### Why Attachments?

- Why are we defining attachments instead of simply adding a field on the KittyVerse NFTs that can store hats?
    - If we didn't define KittyVerse, we couldn't do that - no way we can alter the underlying contract
        - Attachments allow us to permissionlessly extend existing resources with new data and functionality
        - Why would we do that anyway?
            
            - If KittyVerse already has an enthusiastic community behind it
    - Let's say we're defining both, why use attachments?
        - Building with attachments is a clear signal that you're designing for this permissionless composability

>             - You encourage others to define their own sorts of attachments for KittyVerse NFTs

Reviewer: is it possible to create attachments with changing the underlying contract ?

So let's say Alice creates a KittyVerse contract and it doesn't have any attachment functionality and bob owns one kittyVerse and wants to attach a kittyHat NFT to its KittyVerse NFT. Then bob can permisionlessely attach the kittyHat to its kittyVerse without the intervene of Alice changing the contract to add the support of attachments ?

Author: > is it possible to create attachments with changing the underlying contract ?
...
> bob can permisionlessely attach the kittyHat to its kittyVerse without the intervene of Alice changing the contract to add the support of attachments ?

Yes, that's correct. Bob would just create and deploy the KittyHats contract, and then he could add the KittyHats HatAttachment to his KittyVerse NFTs. No change needed by Alice to the KittyVerse contract
